lgc/state/PalMetadata.cpp

```diff
@@ -99,3 +136,126 @@ void PalMetadata::record(Module *module) {
+//                        - one of the UserDataMapping values, e.g. UserDataMapping::GlobalTable
+// @param dwordCount : Number of user data entries to set
+void PalMetadata::setUserDataEntry(ShaderStage stage, unsigned userDataIndex, unsigned userDataValue,
+                                   unsigned dwordCount) {
+  // If this is the spill table pointer, adjust userDataLimit accordingly.
+  if (userDataValue == static_cast<unsigned>(Util::Abi::UserDataMapping::SpillTable))
+    setUserDataLimitToMax();
+
+  // Get the start register number of SPI user data registers for this shader stage.
+  unsigned userDataReg = getUserDataReg0(stage);
+
+  // Assert that the supplied user data index is not too big.
+  assert(userDataIndex + dwordCount <= 32 &&
+         (m_pipelineState->getTargetInfo().getGfxIpVersion().major >= 9 || stage != ShaderStageCompute ||
+          userDataIndex + dwordCount <= 16) &&
+         "Out of range user data index");
+
+  // Update userDataLimit if needed.
+  if (userDataValue < 0x10000 && userDataValue + dwordCount > m_userDataLimit->getUInt())
+    *m_userDataLimit = m_document->getNode(userDataValue + dwordCount);
+
+  // Write the register(s)
+  userDataReg += userDataIndex;
+  while (dwordCount--)
+    m_registers[m_document->getNode(userDataReg++)] = m_document->getNode(userDataValue++);
+}
+
+// =====================================================================================================================
+// Mark that the user data spill table is used at the given offset. The SpillThreshold PAL metadata entry is
+// set to the minimum of any call to this function in any shader.
+//
+// @param dwordOffset : Dword offset that the spill table is used at
+void PalMetadata::setUserDataSpillUsage(unsigned dwordOffset) {
+  if (dwordOffset < m_spillThreshold->getUInt())
+    *m_spillThreshold = m_document->getNode(dwordOffset);
+}
+
+// =====================================================================================================================
+// Finalize PAL metadata for pipeline.
+// TODO Shader compilation: The idea is that this will be called at the end of a pipeline compilation, or in
+// an ELF link, but not at the end of a shader/half-pipeline compile.
+void PalMetadata::finalizePipeline() {
+  // If there are root user data nodes but none of them are used, adjust userDataLimit accordingly.
+  if (m_userDataLimit->getUInt() == 0 && !m_pipelineState->getUserDataNodes().empty())
+    setUserDataLimitToMax();
+}
+
+// =====================================================================================================================
+// Set userDataLimit to maximum (the size of the root user data table, excluding vertex buffer and streamout).
+// This is called if spill is in use, or if there are root user data nodes but none of them are used (PAL does
+// not like userDataLimit being 0 if there are any root user data nodes).
+void PalMetadata::setUserDataLimitToMax() {
```

Comment:
I prefer to use name setUserDataLimit. setUserDataLimitToMax is easy to causes confuse that this function will set userDataLimit to a pre-defined max value.